src/Tasks/FileIO/WriteLinesToFile.cs

```diff
@@ -82,27 +83,31 @@ public override bool Execute()
 
                 try
                 {
+
                     if (Overwrite)
                     {
                         if (buffer.Length == 0)
                         {
-                            // if overwrite==true, and there are no lines to write,
-                            // just delete the file to leave everything tidy.
-                            System.IO.File.Delete(File.ItemSpec);
+                            // File.Delete throws when the directory does not exist.
+                            if (System.IO.Directory.Exists(GetTargetDirectory()))
+                            {
+                                // if overwrite==true, and there are no lines to write,
+                                // just delete the file to leave everything tidy.
+                                System.IO.File.Delete(File.ItemSpec);
+                            }
                         }
                         else
                         {
-                            string contentsAsString = null;
-
+                            CreateTargetDirectory();
```

Comment:
Since `GetTargetDirectory()` is called in both branches, maybe use a local variable before `if (buffer.Length == 0)` which stores `Path.GetDirectoryName(File.ItemSpec)` and use that directly.

Reply:
Not sure I follow correctly. When I place that before `(buffer.Length == 0)` then the `else` branch of the `if (Overwrite)` would not work. I check the directory also when it is appending or shouldn't I do this?

Reply:
See https://github.com/Microsoft/msbuild/pull/4067#discussion_r247311672